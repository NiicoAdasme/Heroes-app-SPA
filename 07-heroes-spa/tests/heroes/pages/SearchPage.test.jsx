import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter, useNavigate } from "react-router-dom"
import { SearchPage } from "../../../src/heroes"

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
}));


describe('Pruebas en el SearchPage', () => {

    test('Debe de mostrar los valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();

    })

    test('Debe de mostrar a batman y el input con el valor del queryString', () => {

        const { container } = render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');

        const img = screen.getByRole('img');
        expect(img.src).toContain('/assets/heroes/dc-batman.jpg')
        expect(container).toMatchSnapshot();

    })

    test('Debe de mostrar un error si no se encuentra el heroe', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batmann']}>
                <SearchPage />
            </MemoryRouter>
        )

        // const alert = screen.getByLabelText('alert-danger');
        // expect( alert ).toBeTruthy();
        // screen.debug();

    })

    test('Debe de llamar el navigate a la ventana nueva', () => {

        const inputValue = 'superman'

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');

        fireEvent.change(input, {target: {name: 'searchText', value: inputValue}})
        console.log(input.value);

        const form = screen.getByRole('form');
        fireEvent.submit(form);

        expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
        
    })

})