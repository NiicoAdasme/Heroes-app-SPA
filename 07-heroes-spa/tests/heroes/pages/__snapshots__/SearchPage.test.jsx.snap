// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el SearchPage Debe de mostrar a batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value="batman"
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Batman"
                class="card-img"
                src="src/assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="car-body"
              >
                <h5
                  class="card-title"
                >
                   
                  Batman
                   
                </h5>
                <p
                  class="card-text"
                >
                   
                  Bruce Wayne
                   
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/heroe/dc-batman"
                >
                  MÃ¡s...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el SearchPage Debe de mostrar los valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero.
      </div>
    </div>
  </div>
</div>
`;
